File: xllm/core/kernels/cuda/activation.cpp
/* Copyright 2025 The xLLM Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://github.com/jd-opensource/xllm/blob/main/LICENSE

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "cuda_ops_api.h"

namespace xllm::kernel::cuda {

void act_and_mul(torch::Tensor out,
                 torch::Tensor input,
                 const std::string& act_mode) {
  if (act_mode != "silu" && act_mode != "gelu" && act_mode != "gelu_tanh") {
    LOG(FATAL) << "Unsupported act mode: " << act_mode
               << ", only support silu, gelu, gelu_tanh";
  }

  std::string uri = act_mode + "_and_mul";

  auto lib =
      torch::DynamicLibrary(path_to_uri_so_lib(uri).c_str(), nullptr, true);
  std::string schema_name = uri + "::" + uri;

  auto act_and_mul_func =
      torch::Dispatcher::singleton()
          .findSchemaOrThrow(schema_name.c_str(), "")
          .typed<void(torch::Tensor&, torch::Tensor&, bool)>();

  act_and_mul_func.call(out, input, support_pdl());
}
}  // namespace xllm::kernel::cuda

File: xllm/python/ops/activation.py
# Copyright 2026 The xLLM Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/xLLM-AI/xllm/blob/main/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Graph interfaces for Python-authored activation operators."""

from __future__ import annotations

import torch


@torch.library.custom_op("xllm_triton::silu_and_mul", mutates_args=())
def silu_and_mul(value: torch.Tensor) -> torch.Tensor:
    """Run the CUDA Triton gated SiLU kernel as one graph node."""
    from xllm.python.kernels.triton.cuda.silu_and_mul import (
        silu_and_mul as triton_silu_and_mul,
    )

    return triton_silu_and_mul(value)


@silu_and_mul.register_fake
def _silu_and_mul_fake(value: torch.Tensor) -> torch.Tensor:
    shape = list(value.shape)
    shape[-1] //= 2
    return value.new_empty(shape)
